Add Password To Redis

Alirexaa opened the pull request:
close: #3838
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4642)

Alirexaa committed:
add password to redis

Alirexaa committed:
Merge branch 'main' into RedisPasswordOptions

eerhardt commented on src/Aspire.Hosting.Redis/PublicAPI.Shipped.txt:
```diff
@@ -7,7 +7,6 @@ Aspire.Hosting.ApplicationModel.RedisResource.RedisResource(string! name) -> voi
 Aspire.Hosting.Redis.RedisCommanderResource
 Aspire.Hosting.Redis.RedisCommanderResource.RedisCommanderResource(string! name) -> void
 Aspire.Hosting.RedisBuilderExtensions
-static Aspire.Hosting.RedisBuilderExtensions.AddRedis(this Aspire.Hosting.IDistributedApplicationBuilder! builder, string! name, int? port = null) -> Aspire.Hosting.ApplicationModel.IResourceBuilder<Aspire.Hosting.ApplicationModel.RedisResource!>!
```
We shouldn't be editing this file. But instead you can add a `*REMOVED*` line to the Unshipped.txt file. See https://github.com/dotnet/aspire/pull/4327/files#diff-6a097377cfc55516be6e941106e3e5575efaa3ab602e773d13d76b803c68c989R25 for an example.

eerhardt commented on src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs:
```diff
@@ -24,13 +24,72 @@ public static class RedisBuilderExtensions
+                              .WithImage(RedisContainerImageTags.Image, RedisContainerImageTags.Tag)
+                              .WithImageRegistry(RedisContainerImageTags.Registry);
+            }
+            throw new InvalidOperationException();
```
Can we add a message to this exception to explain what went wrong?

mitchdenny replied:
It is a shame that we have to do these parameters in this order (its different to `AddPostgres`).

mitchdenny commented on src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs:
```diff
@@ -24,13 +24,72 @@ public static class RedisBuilderExtensions
+    /// <summary>
+    /// Adds a Redis container to the application model.
+    /// </summary>
+    /// <remarks>
```
Could you expand on the remarks here to explain that if a password parameter is not supplied that a password resource will be added to the model a publish time etc. Also as we are trying to improve our docs as we go add in a few `<example>` tags - one for with the password and one for without.

mitchdenny commented:
I'm comfortable with this change pending @eerhardt 's comments about the API txt files and some incremental improvements to the XML docs since we are in the neighbhorhood.

I'm a little disappointed that about the parameter ordering but I can't think of a way that isn't a binary/source breaking change that isn't also disruptive.

One more iteration and we should be able to stamp an approval on this. Thanks once again for your efforts @Alirexaa.

Alirexaa committed:
add xml docs,fix public api text files

Alirexaa committed:
Merge branch 'RedisPasswordOptions' of https://github.com/Alirexaa/aspire into RedisPasswordOptions

radical commented on src/Aspire.Hosting.Redis/RedisResource.cs:
```diff
@@ -18,9 +29,22 @@ public class RedisResource(string name) : ContainerResource(name), IResourceWith
+    /// <summary>
+    /// Gets the parameter that contains the Redis server username.
+    /// </summary>
+
```
```suggestion
```

radical commented on src/Aspire.Hosting.Redis/RedisResource.cs:
```diff
@@ -18,9 +29,22 @@ public class RedisResource(string name) : ContainerResource(name), IResourceWith
+
     private ReferenceExpression ConnectionString =>
+        PasswordParameter is null ?
+        ReferenceExpression.Create(
+            $"{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}") :
         ReferenceExpression.Create(
-            $"{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}");
+            $"{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)},password={PasswordParameter}")
+        ;
```
nit:
```csharp
        PasswordParameter is null
            ? ReferenceExpression.Create(
                $"{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}")
            : ReferenceExpression.Create(
                $"{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)},password={PasswordParameter}");
```

radical commented on tests/Aspire.Hosting.Tests/Redis/AddRedisTests.cs:
```diff
@@ -69,7 +69,24 @@ public void AddRedisContainerAddsAnnotationMetadata()
+        var connectionStringResource = Assert.Single(appModel.Resources.OfType<IResourceWithConnectionString>());
+        Assert.Equal("{myRedis.bindings.tcp.host}:{myRedis.bindings.tcp.port},password={pass.value}", connectionStringResource.ConnectionStringExpression.ValueExpression);
+    }
+    [Fact]
```
```suggestion

    [Fact]
```

radical commented on tests/Aspire.Hosting.Tests/Redis/AddRedisTests.cs:
```diff
@@ -69,7 +69,24 @@ public void AddRedisContainerAddsAnnotationMetadata()
+        Assert.Equal("{myRedis.bindings.tcp.host}:{myRedis.bindings.tcp.port},password={pass.value}", connectionStringResource.ConnectionStringExpression.ValueExpression);
+    }
+    [Fact]
+    public async Task RedisCreatesConnectionStringWithPassword()
```
Should this be `*WithoutPassword`?
```suggestion
    public async Task RedisCreatesConnectionStringWithoutPassword()
```

radical commented on tests/Aspire.Hosting.Tests/Redis/AddRedisTests.cs:
```diff
@@ -69,7 +69,24 @@ public void AddRedisContainerAddsAnnotationMetadata()
 
     [Fact]
-    public async Task RedisCreatesConnectionString()
+    public void RedisCreatesConnectionString()
```
```suggestion
    public void RedisCreatesConnectionStringWithPassword()
```

radical commented on tests/Aspire.Hosting.Tests/Redis/AddRedisTests.cs:
```diff
@@ -341,4 +394,29 @@ public void WithPersistenceAddsCommandLineArgsAnnotation()
         Assert.True(redis.Resource.TryGetAnnotationsOfType<CommandLineArgsCallbackAnnotation>(out var argsAnnotations));
         Assert.NotNull(argsAnnotations.SingleOrDefault());
     }
+    [Fact]
```
```suggestion

    [Fact]
```

radical commented on src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs:
```diff
@@ -19,18 +19,129 @@ public static class RedisBuilderExtensions
+    /// </summary>
+    /// <remarks>
+    /// The default image is "redis" and the tag is "7.2.4".
+    /// If a password parameter is not supplied, a password resource will be added to the model at publish time.
```
Just curious - what password would be used in this case? And does that conflict with `password">The parameter used to provide the password for the Redis resource. If <see langword="null"/> resource will be not password protected`

radical replied:
Maybe we also need a test to check that if the password parameter is not supplied then a password resource is being added to the model at publish time?

radical commented on tests/Aspire.Hosting.Tests/Redis/AddRedisTests.cs:
```diff
@@ -111,6 +128,40 @@ public async Task VerifyManifest()
     }
 
+    [Fact]
+    public async Task VerifyWithPasswordManifest()
```
Do we also need tests for when `builder.Configuration["Parameters:pass"] = password` is missing?

radical commented on tests/Aspire.Hosting.Tests/Redis/AddRedisTests.cs:
```diff
@@ -69,7 +69,24 @@ public void AddRedisContainerAddsAnnotationMetadata()
+        appBuilder.Configuration["Parameters:pass"] = password;
+
+        var pass = appBuilder.AddParameter("pass");
+        appBuilder.AddRedis("myRedis", password: pass).PublishAsContainer();
```
Maybe another test for `AddRedis("myredis", port: .., password: ..)`?

radical commented:
Some of my comments might be non-sensical, or could be addressed in a follow-up PR. I'll let others decide that, since I'm very new to this code:)

eerhardt commented on playground/Redis/Redis.AppHost/Program.cs:
```diff
@@ -0,0 +1,20 @@
+builder.Configuration["Parameters:pass"] = "123456";
+var password = builder.AddParameter("pass");
+
+var redis = builder.AddRedis("redis", password: password);
```
Can we `WithRedisCommander` here too?

DamianEdwards commented on src/Aspire.Hosting.Redis/RedisResource.cs:
```diff
@@ -18,9 +29,22 @@ public class RedisResource(string name) : ContainerResource(name), IResourceWith
+            $"{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}") :
         ReferenceExpression.Create(
-            $"{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}");
+            $"{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)},password={PasswordParameter}")
```
Does the password need to be encoded here, e.g. URI encoded?

Alirexaa committed:
Update src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

Alirexaa committed:
Update src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

Alirexaa committed:
Update src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

Alirexaa committed:
Apply suggestions from code review

Co-authored-by: Ankit Jain <radical@gmail.com>

Alirexaa committed:
apply suggested changes

Alirexaa committed:
add more tests

Alirexaa committed:
add password to redis commander REDIS_HOSTS env variable

Alirexaa commented on src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs:
```diff
@@ -19,18 +19,129 @@ public static class RedisBuilderExtensions
+        if (password?.Resource is not null)
+        {
+
+            var redis = new RedisResource(name, password.Resource);
```
How?

mgravell replied:
We've discussed this for the client a few times but it has never hit the top of the pile. If it is a blocker, we can prioritize. I would suggest looking at ado-net quoted values as the template, but obvs we'd need to think a little about how to do it without impacting existing use of any new token

eerhardt replied:
Are you able to write a test?

Alirexaa replied:
Do you mean Writing e2e test?

I think RedisCommander start failed when Redis have password but password was not provided in connectionString (here REDIS_HOSTS environments variable).

I could check RedisCommander start when Redis have password, Right?

eerhardt replied:
Yes, I think writing a test that has:

* a Redis resource
* a RedisCommander resource

And running it, ensuring that the RedisCommander resource starts without any errors.

Alirexaa replied:
@eerhardt, please take a look at the written test.
Also, it needs to import the RedisCommander image to `netaspireci.azurecr.io` to work

eerhardt replied:
Thanks. I've imported the RedisCommander image to netaspireci.azurecr.io.

mgravell commented:
@eerhardt We've discussed this for the client a few times but it has never hit the top of the pile. If it is a blocker, we can prioritize. I would suggest looking at ado-net quoted values as the template, but obvs we'd need to think a little about how to do it without impacting existing use of any new token

eerhardt committed:
Minor PR feedback

eerhardt commented:
> @eerhardt We've discussed this for the client a few times but it has never hit the top of the pile. If it is a blocker, we can prioritize. I would suggest looking at ado-net quoted values as the template, but obvs we'd need to think a little about how to do it without impacting existing use of any new token

I don't think it is blocking this work. I just wanted to point out that we can't URL encode the connection string here.

eerhardt commented on src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs:
```diff
@@ -19,18 +19,124 @@ public static class RedisBuilderExtensions
+        int? port = null,
+        IResourceBuilder<ParameterResource>? password = null)
+    {
+        if (password?.Resource is not null)
```
@DamianEdwards @mitchdenny @davidfowl - what do you think about not having this logic and just always adding a password if it is null here? Yes, it is technically a behavior breaking change. But it will make Redis consistent with the others that have passwords - mysql, postgres, etc.

mitchdenny replied:
I agree let's shoot for consistency here.

Alirexaa replied:
so, we don't need to check the execution context, always adding a password to the resource whether the app is in publish mode or run mode. right?

mitchdenny replied:
If it is in publish mode and the password isn't provided we'll create a synthetic parameter like we do for Postgres.

Alirexaa replied:
we already did. look at this

```
        if (builder.ExecutionContext.IsPublishMode)
        {
            var passwordParameter = password?.Resource ?? ParameterResourceBuilderExtensions.CreateDefaultPasswordParameter(builder, $"{name}-password");
            var redis = new RedisResource(name, passwordParameter);

            return builder.AddResource(redis)
                      .WithEndpoint(port: port, targetPort: 6379, name: RedisResource.PrimaryEndpointName)
                      .WithImage(RedisContainerImageTags.Image, RedisContainerImageTags.Tag)
                      .WithImageRegistry(RedisContainerImageTags.Registry);
                      .WithImageRegistry(RedisContainerImageTags.Registry)
                      .WithPassword(passwordParameter);
        }
        else if (builder.ExecutionContext.IsRunMode)
        {
            var redis = new RedisResource(name);
            return builder.AddResource(redis)
                          .WithEndpoint(port: port, targetPort: 6379, name: RedisResource.PrimaryEndpointName)
                          .WithImage(RedisContainerImageTags.Image, RedisContainerImageTags.Tag)
                          .WithImageRegistry(RedisContainerImageTags.Registry);
        }
```

eerhardt replied:
I've push an update for what I was talking about. We always add a password when calling `.AddRedis("name")` now. Just like other resources like PostgreSQL.

Alirexaa committed:
Merge branch 'main' into RedisPasswordOptions

Alirexaa committed:
fix failing test

Alirexaa committed:
fix confilict from main to RedisPasswordOptions

Alirexaa committed:
add test for redis commander shuold works when password provided

radical commented on src/Aspire.Hosting.Redis/RedisCommanderConfigWriterHook.cs:
```diff
@@ -31,7 +31,15 @@ public Task AfterEndpointsAllocatedAsync(DistributedApplicationModel appModel, C
+                }
+                else
+                {
+                    hostString = $"{(hostsVariableBuilder.Length > 0 ? "," : string.Empty)}{redisInstance.Name}:{redisInstance.PrimaryEndpoint.ContainerHost}:{redisInstance.PrimaryEndpoint.Port}:0";
```
Test needed for this case.

Alirexaa replied:
Isn't these enough?

https://github.com/dotnet/aspire/blob/43e1cf4bb6493848c7fc60d15fe565677e69abda/tests/Aspire.Hosting.Redis.Tests/AddRedisTests.cs#L159-L180

https://github.com/dotnet/aspire/blob/43e1cf4bb6493848c7fc60d15fe565677e69abda/tests/Aspire.Hosting.Redis.Tests/AddRedisTests.cs#L182-L208

radical replied:
Do these cover the `else` block of no password available?

Alirexaa replied:
Could you please explain the exact senario you mean?
For me it seems cover.

radical replied:
Maybe I'm misunderstanding something. This was prompted by looking at the code coverage report for that file. And looking at the code my understanding is that the test has `builder.AddRedis("myredis1")` and does not add a password explicitly, so we generate and add one by default. If that's the case then would this line of code be hit, as it is would run `if (redisInstance.PasswordParameter is not null)` is false.

Please correct me if I misunderstood the code.

<img width="1275" alt="Screenshot 2024-07-11 at 19 24 16" src="https://github.com/user-attachments/assets/13c99752-213b-47a7-bbda-ed861731e18a">

Alirexaa replied:
in these two mentioned cases, we use the `Run` mode so the password is null and we do not add the password by default. so else branch be hit.
for just now (8.1), we generate a password by default when the execution context is `Publish` and the password was not added explicitly.

I think we just need a test case when a password was not provided and the execution context is `Publish`.

Alirexaa replied:
> I think we just need a test case when a password was not provided and the execution context is Publish.

@radical, there is no need to test for this situation, because RedisCommander be excluded from the manifest.

https://github.com/dotnet/aspire/blob/cf61a58a739f059469bcbfbff7d5121526f6fa88/src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs#L56-L61

Alirexaa commented:
What is the status of PR?

Currntly i'm working on #5165,
I need playground app there,
Besides that, if redis have password, i need to support password there.

@eerhardt @mitchdenny

mitchdenny commented:
@Alirexaa see: https://github.com/dotnet/aspire/pull/4642#discussion_r1669073634

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into RedisPasswordOptions

eerhardt committed:
PR feedback

- Update for latest code.
- Always generate a password parameter, like the other resources.
- Handle persistence and password command line annotation conflicts by having a single commandline callback annotation, and combine them both together.
- Update tests

eerhardt committed:
Remove playground build files.

eerhardt committed:
Remove explicit dashboard reference.

mitchdenny commented:
I've been talking a closer look at this from a security perspective, and I think we need to make some changes. When I run this locally the args are visble in the dashboard without any obfuscation (unlike environment variables).

I'm looking at the implications for deployment now too.

Alirexaa committed:
Merge branch 'dotnet:main' into RedisPasswordOptions

Alirexaa committed:
fix test

eerhardt commented:
> I've been talking a closer look at this from a security perspective, and I think we need to make some changes. When I run this locally the args are visble in the dashboard without any obfuscation (unlike environment variables).
> 
> I'm looking at the implications for deployment now too.
> 
> OK, so deployment doesn't use the args which means that the password isn't required. So we are going to need to fix that in `azd` before this change would have any impact (it has a local impact obviously but we need to think through to deployment for this one).

Good catch.

Reading https://redis.io/docs/latest/operate/oss_and_stack/management/security/#authentication, the other options for enabling a password are:

1. Adding a user and password to the [Access Control List](https://redis.io/docs/latest/operate/oss_and_stack/management/security/acl/)
2. Use the `requirepass` setting in the redis.conf file.

Either approach would require being able to write a file that is mounted into the container. Is that possible during deployment? @ellismg @vhvb1989

Alternatively, we could modify the Access Control List by connecting to the redis container after it is running, and issuing commands to it. I don't believe this is possible during deployment.

eerhardt commented:
>  When I run this locally the args are visble in the dashboard without any obfuscation (unlike environment variables).

@drewnoakes @adamint - is it possible to mark command line arguments as "secret"? So they get obfuscated in the dashboard.

vhvb1989 commented:
> Either approach would require being able to write a file that is mounted into the container. Is that possible during deployment?

That's `bindMounts`, isn't it?
Like the sample from MySql [here](https://github.com/dotnet/aspire/blob/main/playground/mysql/MySqlDb.AppHost/Program.cs#L9)

The only difference would be that the file that is uploaded is generated, instead of committing it to the repo (or it can be a file in the repo with references to secured-args)

We support bind mount with an alpha feature in azd.  During provision, azd creates an storage account and one file-share which is set up as volume in the Azure Container Environment.  After provision, azd uploads the local content to the file-share. At the end, azd deploys the Azure Container App using the volume

drewnoakes commented:
> @drewnoakes @adamint - is it possible to mark command line arguments as "secret"? So they get obfuscated in the dashboard.

#5065 tracks marking resource properties as secrets, however that has to be specified by the resource service. We could hard-code some known properties as sensitive. If we did that, which would you think should be included? The current list is https://github.com/dotnet/aspire/blob/0f2931a94873d09e5a76f72485000e5388bf08dd/src/Shared/Model/KnownProperties.cs

mitchdenny commented:
As much as I would like this in for 8.2, I think we need to hold this PR until we get this resolved. I think hiding the args in the dashboard is fine for local given that if you have access to the dashboard you could uncover the secrets in environment variables anyway.

I'm more concerned about deployment where people will have a reasonable expectation that the if you specify the password that it'll be used when deployed to the cloud.

Frustratingly Redis Stack actually makes use of a `REDIS_ARGS` environment variable which can use to specify the `-requirepass [password]` argument.

eerhardt commented:
> Frustratingly Redis Stack actually makes use of a REDIS_ARGS environment variable which can use to specify the -requirepass [password] argument.

Yeah, I was wondering if it would make sense for us to switch from `library/redis` to `redis/redis-stack-server` instead. And then [Add WithRedisInsight (dotnet/aspire#5227)](https://github.com/dotnet/aspire/pull/5227) would switch the image from `redis/redis-stack-server` to `redis/redis-stack` and add the appropriate endpoints.

drewnoakes commented:
> > @drewnoakes @adamint - is it possible to mark command line arguments as "secret"? So they get obfuscated in the dashboard.
> 
> #5065 tracks marking resource properties as secrets, however that has to be specified by the resource service. We could hard-code some known properties as sensitive. If we did that, which would you think should be included? The current list is https://github.com/dotnet/aspire/blob/0f2931a94873d09e5a76f72485000e5388bf08dd/src/Shared/Model/KnownProperties.cs

This is in review in https://github.com/dotnet/aspire/pull/5380

mitchdenny commented:
> Yeah, I was wondering if it would make sense for us to switch from `library/redis` to `redis/redis-stack-server` instead. And then [Add WithRedisInsight (dotnet/aspire#5227)](https://github.com/dotnet/aspire/pull/5227) would switch the image from `redis/redis-stack-server` to `redis/redis-stack` and add the appropriate endpoints.

I think this is an option we should seriously consider. I think we would just need to look through where there are any licensing implications for folks deploying that image into production. There might be extra licensing requirements that we wouldn't want to silently swap people into.

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into RedisPasswordOptions

eerhardt committed:
Fix up merge for the latest code

eerhardt committed:
Fix Redis Insights with a Redis connection with a password.

This needs to work around https://github.com/RedisInsight/RedisInsight/issues/3452, which requires the agreements encryption property to be initialized.

eerhardt commented:
I've revived this PR with the latest code in `main`. The intent is to get this change into 9.1. The [initial concerns](https://github.com/dotnet/aspire/pull/4642#issuecomment-2292514163) that blocked this change should no longer be a problem. The command line args in the dashboard are hidden/masked by default now.

One complication is that Redis Insight needs to be initialized before we add a database with a password. I'll add a comment to the code to start a conversation about it there.

eerhardt commented on src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs:
```diff
@@ -305,10 +379,55 @@ await pipeline.ExecuteAsync(async (ctx) =>
+            agreements = new
+            {
+                // all 4 are required to be set
+                eula = false,
```
This is unfortunate, but it is the only way I could get it to work. See https://github.com/RedisInsight/RedisInsight/issues/3452 for more information on why this is needed.

Note that this isn't accepting the EULA (it is sending false). But the consequences of initializing this is the pop up no longer happens on first run of Redis Insight, which is a nice user experience.

Thoughts? @sebastienros @DamianEdwards @Alirexaa

davidfowl replied:
What are the options?

DamianEdwards replied:
Seems [they're suggesting we wait](https://github.com/RedisInsight/RedisInsight/issues/3452#issuecomment-2623640956) until they implement support for adding databases via environment variables instead.

eerhardt replied:
I've opened [[Bug]: Posting eula: false to /api/settings suppresses the first run dialog (RedisInsight/RedisInsight#4340)](https://github.com/RedisInsight/RedisInsight/issues/4340) for the first run dialog issue.

Let's move forward with this PR for now. We can readjust later.

eerhardt committed:
- Fix tests
- Update manifests

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into RedisPasswordOptions

eerhardt committed:
Fix new test.

davidfowl commented on src/Aspire.Hosting.Redis/RedisResource.cs:
```diff
@@ -18,9 +28,20 @@ public class RedisResource(string name) : ContainerResource(name), IResourceWith
+    /// <summary>
+    /// Gets the parameter that contains the Redis server username.
+    /// </summary>
     private ReferenceExpression ConnectionString =>
```
Prefer this pattern https://github.com/dotnet/aspire/blob/9e2a2dfbeda1b85501b3f80ad58f50f4609a1a09/src/Aspire.Hosting.MongoDB/MongoDBServerResource.cs#L56

davidfowl reviewed (approved):
Do we mark this as a breaking change?

davidfowl committed:
Merge branch 'main' into RedisPasswordOptions

eerhardt commented:
> is it possible to specify the user for Redis? Or is that unnecessary in this scenario, with Aspire just using the default user?

Right now, no. We are just adding password protection for now. If necessary, we can also add support for user in the future.

> Similarly, there is increasing support for azure etc identity tokens for auth.

We support Azure Redis and Entra ID already in 9.0. See https://learn.microsoft.com/dotnet/aspire/caching/stackexchange-redis-integration#add-azure-cache-for-redis-client

eerhardt commented:
> Do we mark this as a breaking change?

I don't think this is a breaking change. We can log it in the changelog, but I'm not sure exactly what it would break. The only thing I can think of is if someone is explicitly cutting the password out of the connection string getting flown to the app (or passing the connection string a custom way - not using our Hosting ConnectionString).

eerhardt commented:
Thank you, @Alirexaa for your continued contributions!
